### PR TITLE
Remove unused errors

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
* Removes errors that aren't used by the compiler (mostly duplicates)
* Adds a script that finds these + finds errors that aren't anywhere in the baselines (in case we get bored some day and need something to do...)

### CODE REVIEW

File: scripts/errorCheck.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
+declare var require: any;
+let fs = require('fs');
+let async = require('async');
+let glob = require('glob');
+
+fs.readFile('src/compiler/diagnosticMessages.json', 'utf-8', (err, data) => {
+	let messages  = JSON.parse(data);
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: This file uses tabs, switch to spaces

File: scripts/errorCheck.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
+declare var require: any;
+let fs = require('fs');
+let async = require('async');
+let glob = require('glob');
+
+fs.readFile('src/compiler/diagnosticMessages.json', 'utf-8', (err, data) => {
+	let messages  = JSON.parse(data);
+	if (err) throw err;
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: curlies

File: scripts/errorCheck.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+let async = require('async');
+let glob = require('glob');
+
+fs.readFile('src/compiler/diagnosticMessages.json', 'utf-8', (err, data) => {
+	let messages  = JSON.parse(data);
+	if (err) throw err;
+	let keys = Object.keys(messages);
+	console.log('Loaded ' + keys.length + ' errors');
+
+	for(let k of keys) {
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: space after `for`

File: scripts/errorCheck.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	let errRegex = /\(\d+,\d+\): error TS([^:]+):/g;
+
+	let baseDir = 'tests/baselines/reference/';
+	fs.readdir(baseDir, (err, files) => {
+		files = files.filter(f => f.indexOf('.errors.txt') > 0);
+		console.log('Read ' + files.length + ' baselines');
+		let tasks: Array<(callback: () => void) => void> = [];
+		files.forEach(f => tasks.push(done => {
+			fs.readFile(baseDir + f, 'utf-8', (err, baseline) => {
+				if (err) throw err;
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: curlies

File: scripts/errorCheck.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+				let g: string[];
+				while(g = errRegex.exec(baseline)) {
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: space after `while`

File: scripts/errorCheck.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+					var errCode = +g[1];
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: Use `let`

File: scripts/errorCheck.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+				done();
+			});
+		}));
+
+		async.parallelLimit(tasks, 25, done => {
+			console.log('== List of errors not present in baselines ==');
+			let count = 0;
+			for(let k of keys) {
+				if(messages[k]['seen'] !== true) {
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: Space after `if` and `for`

File: scripts/errorCheck.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+				}
+
+				done();
+			});
+		}));
+
+		async.parallelLimit(tasks, 25, done => {
+			console.log('== List of errors not present in baselines ==');
+			let count = 0;
+			for(let k of keys) {
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: You could use `keys.forEach` since it calls back with an index

File: scripts/errorCheck.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+			console.log(count + ' of ' + keys.length + ' errors are not in baselines');
+		});
+	});
+});
+
+fs.readFile('src/compiler/diagnosticInformationMap.generated.ts', 'utf-8', (err, data) => {
+	let errorRegexp = /\s(\w+): \{ code/g;
+	let errorNames: string[] = [];
+	let errMatch: string[];
+	while(errMatch = errorRegexp.exec(data)) {
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: space after `while`

File: scripts/errorCheck.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+		errorNames.push(errMatch[1]);
+	}
+
+	let allSrc: string = '';
+	glob('./src/**/*.ts', {}, (err, files) => {
+		console.log('Reading ' + files.length + ' source files');
+		files.forEach(file => {
+			if (file.indexOf('diagnosticInformationMap.generated.ts') > 0) return;
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: Curlies

File: scripts/errorCheck.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+			let src = fs.readFileSync(file, 'utf-8');
+			allSrc = allSrc + src;
+		});
+
+		console.log('Consumed ' + allSrc.length + ' characters of source');
+
+		let count = 0;
+		console.log('== List of errors not used in source ==')
+		errorNames.forEach(errName => {
```
Comment (DanielRosenwasser) [inline scripts/errorCheck.ts]: for-of could work here

File: src/compiler/diagnosticMessages.json [hunk 12]
```diff
@@ -67,22 +67,6 @@
...
-        "code": 1025
-    },
-    "A class can only extend a single class.": {
-        "category": "Error",
-        "code": 1026
-    },
-    "A class declaration can only have one 'implements' clause.": {
-        "category": "Error",
-        "code": 1027
-    },
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: These seem like legitimate errors. What do we report in place of them?

Comment (RyanCavanaugh) [inline src/compiler/diagnosticMessages.json]:
```ts
extends_clause_already_seen
extends_clause_must_precede_implements_clause
Classes_can_only_extend_a_single_class
implements_clause_already_seen
```

Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: Those seem like potentially better messages given the direction of singular indefinite articles @JsonFreeman pushed for.

Comment (RyanCavanaugh) [inline src/compiler/diagnosticMessages.json]: These are in line with our other parser-like error messages, e.g. we say `"Unterminated regular expression literal."` not `A regular expression literal must be terminated`.

File: scripts/errorCheck.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+		async.parallelLimit(tasks, 25, done => {
+			console.log('== List of errors not present in baselines ==');
+			let count = 0;
+			for (let k of keys) {
+				if (messages[k]['seen'] !== true) {
+					console.log(k);
+					count++;
+				}
+			}
+			console.log(count + ' of ' + keys.length + ' errors are not in baselines');
```
Comment (mhegazy) [inline scripts/errorCheck.ts]: What are the numbers your found?

**Comment by mhegazy** [comment]
:+1:

File: scripts/errorCheck.ts (hunk 13 above)
Comment (RyanCavanaugh) [inline scripts/errorCheck.ts]: 138 of 535 errors are not in baselines

Comment (RyanCavanaugh) [inline scripts/errorCheck.ts]: Note that a decent number of those are e.g. commandline option descriptions, which aren't actually errors

**Comment by yuit** [comment]
:+1:

**Pull request merged**